### ISSUE TITLE
[JENKINS-27708] Concurrent build limits not honored on Jenkins 1.607

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#513]
<p>After upgrading to 1.607 we noticed the throttle plugin doesn't always prevent jobs from running in parallel as expected.</p>

<p><b>Steps to reproduce:</b><br/>
1. Create a throttle category with global limit 1, per-node limit 1.<br/>
2. Create 3 jobs using the category, restricted to a node with two executors, sleep 60 as a build step.<br/>
3. Request builds of all 3 jobs.</p>

<p><b>What should happen:</b><br/>
4. Jobs run in sequence; 1 then 2 then 3.</p>

<p><b>What actually happens:</b><br/>
5. Job 1 starts building, jobs 2 and 3 wait in queue (OK).<br/>
6. After job 1 finishes, both job 2 and 3 start running (<b>not OK</b>).</p>

<p>Plugin version is 1.8.4. Issue does not appear in Jenkins 1.606 with the same version of the plugin. Issue is reproducible on a fresh install.</p>

<p>We were forced to downgrade back to 1.606 as a workaround (which is unfortunately not trivial due to <a href="https://issues.jenkins.io/browse/JENKINS-27700" title="Node configuration missing after downgrading from 1.607 to 1.606" class="issue-link" data-issue-key="JENKINS-27700"><del>JENKINS-27700</del></a>).</p>

---
<details><summary><i>Originally reported by <a title="Tomasz Śniatowski" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tsniatowski">tsniatowski</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-27708" target="_blank">Concurrent build limits not honored on Jenkins 1.607</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Oleg Nenashev" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>
<li><b>reported by</b>: tsniatowski
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-04-21T17:33:21+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 17
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://github.com/jenkinsci/jenkins/pull/1596" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/1596</a> reworks thread synchronization approaches in the core. It may cause such issue.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly">stephenconnolly</a>:</i>
<p>Actually I suspect that <a href="https://github.com/jenkinsci/jenkins/pull/1596" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/1596</a> just exposes problems with throttle concurrent builds plugin that were masked.</p>

<p>Throttle concurrent builds "works" by searching for jobs that are running during the QueueTaskDispatcher.canTake evaluation. In Jenkins core, pre: <a href="https://github.com/jenkinsci/jenkins/blob/jenkins-1.606/core/src/main/java/hudson/model/Queue.java#L1178" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/jenkins-1.606/core/src/main/java/hudson/model/Queue.java#L1178</a> and post: <a href="https://github.com/jenkinsci/jenkins/blob/jenkins-607/core/src/main/java/hudson/model/Queue.java#L1352" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/jenkins-607/core/src/main/java/hudson/model/Queue.java#L1352</a> the QueueTaskDispatcher methods are called for all Buildable jobs in the Queue <b>before</b> mapping any of them against an executor.</p>

<p>Thus <a href="https://github.com/jenkinsci/throttle-concurrent-builds-plugin/blob/throttle-concurrents-1.8.4/src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java#L194" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/throttle-concurrent-builds-plugin/blob/throttle-concurrents-1.8.4/src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java#L194</a> will always fail to count jobs that are being picked off the queue at the same time.</p>

<p>What the TCB plugin should probably be doing is either:</p>

<ul>
	<li>using ResourceLists <a href="https://github.com/jenkinsci/jenkins/blob/jenkins-1.606/core/src/main/java/hudson/model/ResourceController.java#L80" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/jenkins-1.606/core/src/main/java/hudson/model/ResourceController.java#L80</a> to prevent the build from actually starting or checking against both projects on</li>
</ul>


<ul>
	<li>or when counting builds on a node: <a href="https://github.com/jenkinsci/throttle-concurrent-builds-plugin/blob/throttle-concurrents-1.8.4/src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java#L194" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/throttle-concurrent-builds-plugin/blob/throttle-concurrents-1.8.4/src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java#L194</a>  it should actually be including also the Buildable projects against that computer <a href="https://github.com/jenkinsci/jenkins/blob/jenkins-1.607/core/src/main/java/hudson/model/Queue.java#L846" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/blob/jenkins-1.607/core/src/main/java/hudson/model/Queue.java#L846</a></li>
</ul>


<p>The code change in 1.607 just removed some of the threading issues that allowed the TCB plugin to work most of the time <b>by accident</b> this bug represents a logical flaw in TCB that could still occur in 1.536-1.606 (most likely if running on a single core master) if the thread sequencing falls just right as Queue.maintain evaluates all projects in Buildable in one chunk and then maps them against executors. </p>

<p>With 1.536-1.606 the executor threads could pick up the tasks while the Queue was in the process of dividing out the atomic set of work assignments, and thus - as long as the executor threads were scheduled fast enough - the TCB plugin would see the work assignments in the process of being assigned (but on a heavily loaded system or a single core system this issue would still occur)</p>

<p>With 1.607 we ensure that the work assignment from Queue.maintain is mapped as an atomic unit, and thus the bug in TCB is exposed always.</p>

<p>At least that is my analysis </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=zowers">zowers</a>:</i>
<p>> It seems to be a showstopper for jenkins-1.607+.<br/>
the issue's priority should be changed to blocker then</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly">stephenconnolly</a>:</i>
<p>Here is one final version of a fix, presented for posterity - or if we need to revisit. </p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">diff --git a/core/src/main/java/hudson/model/Queue.java b/core/src/main/java/hudson/model/Queue.java
index d776a57..67ba711 100644
--- a/core/src/main/java/hudson/model/Queue.java
+++ b/core/src/main/java/hudson/model/Queue.java
@@ -340,6 +340,25 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
 
     <span class="code-keyword">private</span> <span class="code-keyword">transient</span> <span class="code-keyword">final</span> Condition condition = lock.newCondition();
 
+    /**
+     * There are some plugins that determine whether a project is blocked from execution on the basis of the 
+     * jobs currently executing or currently pending execution. When we ask <span class="code-keyword">if</span> a {@link Queue.Task} is blocked
+     * <span class="code-keyword">for</span> execution from {@link #maintain()}, we need to ensure that the {@link Queue.Task#isBuildBlocked(Item)}
+     * is returning a result based on the live information of jobs scheduled rather than a result based on the
+     * most recent {@link #snapshot}. This field holds the identity of the current thread that is running 
+     * {@link #maintain()} and thus we can <span class="code-keyword">switch</span> to returning live results when called from that thread.
+     * 
+     * Two alternative fixes <span class="code-keyword">for</span> <span class="code-keyword">this</span> <span class="code-keyword">class </span>of issue (JENKINS-22708 and JENKINS-27871) could also be used:
+     * <ul>
+     *     <li>Update the {@link #snapshot}</li> after each and every {@link Queue.Task} is assigned <span class="code-keyword">for</span> execution.</li>
+     *     <li>Have {@link #maintain()} <span class="code-keyword">return</span> after finding one {@link Queue.Task} to execute.</li>
+     * </ul>
+     * @since 1.FIXME
+     * @see #liveResultRequired() 
+     */
+    @CheckForNull
+    <span class="code-keyword">private</span> <span class="code-keyword">transient</span> <span class="code-keyword">volatile</span> <span class="code-object">Thread</span> maintainThread = <span class="code-keyword">null</span>;
+
     <span class="code-keyword">public</span> Queue(@Nonnull LoadBalancer loadBalancer) {
 <span class="code-keyword">this</span>.loadBalancer =  loadBalancer.sanitize();
 <span class="code-comment">// <span class="code-keyword">if</span> all the executors are busy doing something, then the queue won't be maintained in
</span>@@ -844,6 +863,19 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
      * Gets all the {@link BuildableItem}s that are waiting <span class="code-keyword">for</span> an executor in the given {@link Computer}.
      */
     <span class="code-keyword">public</span> List<BuildableItem> getBuildableItems(Computer c) {
+<span class="code-keyword">if</span> (liveResultRequired() && lock.tryLock()) {
+    <span class="code-comment">// JENKINS-22708 and JENKINS-27871. Note we should get the lock always as it should be already held
</span>+    <span class="code-comment">// but the safer code construct is to get the lock anyway rather than assume it is held just by virtue
</span>+    <span class="code-comment">// of the current thread being the thread running maintain().
</span>+    <span class="code-keyword">try</span> {
+List<BuildableItem> result = <span class="code-keyword">new</span> ArrayList<BuildableItem>();
+_getBuildableItems(c, buildables, result);
+_getBuildableItems(c, pendings, result);
+<span class="code-keyword">return</span> result;
+    } <span class="code-keyword">finally</span> {
+lock.unlock();
+    }
+}
 Snapshot snapshot = <span class="code-keyword">this</span>.snapshot;
 List<BuildableItem> result = <span class="code-keyword">new</span> ArrayList<BuildableItem>();
 _getBuildableItems(c, snapshot.buildables, result);
@@ -865,6 +897,18 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
      * Gets the snapshot of all {@link BuildableItem}s.
      */
     <span class="code-keyword">public</span> List<BuildableItem> getBuildableItems() {
+<span class="code-keyword">if</span> (liveResultRequired() && lock.tryLock()) {
+    <span class="code-comment">// JENKINS-22708 and JENKINS-27871. Note we should get the lock always as it should be already held
</span>+    <span class="code-comment">// but the safer code construct is to get the lock anyway rather than assume it is held just by virtue
</span>+    <span class="code-comment">// of the current thread being the thread running maintain().
</span>+    <span class="code-keyword">try</span> {
+ArrayList<BuildableItem> r = <span class="code-keyword">new</span> ArrayList<BuildableItem>(buildables);
+r.addAll(pendings);
+<span class="code-keyword">return</span> r;
+    } <span class="code-keyword">finally</span> {
+lock.unlock();
+    }
+}
 Snapshot snapshot = <span class="code-keyword">this</span>.snapshot;
 ArrayList<BuildableItem> r = <span class="code-keyword">new</span> ArrayList<BuildableItem>(snapshot.buildables);
 r.addAll(snapshot.pendings);
@@ -875,6 +919,16 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
      * Gets the snapshot of all {@link BuildableItem}s.
      */
     <span class="code-keyword">public</span> List<BuildableItem> getPendingItems() {
+<span class="code-keyword">if</span> (liveResultRequired() && lock.tryLock()) {
+    <span class="code-comment">// JENKINS-22708 and JENKINS-27871. Note we should get the lock always as it should be already held
</span>+    <span class="code-comment">// but the safer code construct is to get the lock anyway rather than assume it is held just by virtue
</span>+    <span class="code-comment">// of the current thread being the thread running maintain().
</span>+    <span class="code-keyword">try</span> {
+<span class="code-keyword">return</span> <span class="code-keyword">new</span> ArrayList<BuildableItem>(pendings);
+    } <span class="code-keyword">finally</span> {
+lock.unlock();
+    }
+}
 <span class="code-keyword">return</span> <span class="code-keyword">new</span> ArrayList<BuildableItem>(snapshot.pendings);
     }
 
@@ -902,6 +956,21 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
      * @since 1.402
      */
     <span class="code-keyword">public</span> List<Item> getUnblockedItems() {
+<span class="code-keyword">if</span> (liveResultRequired() && lock.tryLock()) {
+    <span class="code-comment">// JENKINS-22708 and JENKINS-27871. Note we should get the lock always as it should be already held
</span>+    <span class="code-comment">// but the safer code construct is to get the lock anyway rather than assume it is held just by virtue
</span>+    <span class="code-comment">// of the current thread being the thread running maintain().
</span>+    <span class="code-keyword">try</span> {
+List<Item> queuedNotBlocked = <span class="code-keyword">new</span> ArrayList<Item>();
+queuedNotBlocked.addAll(waitingList);
+queuedNotBlocked.addAll(buildables);
+queuedNotBlocked.addAll(pendings);
+<span class="code-comment">// but not <span class="code-quote">'blockedProjects'</span>
</span>+<span class="code-keyword">return</span> queuedNotBlocked;
+    } <span class="code-keyword">finally</span> {
+lock.unlock();
+    }
+}
 Snapshot snapshot = <span class="code-keyword">this</span>.snapshot;
 List<Item> queuedNotBlocked = <span class="code-keyword">new</span> ArrayList<Item>();
 queuedNotBlocked.addAll(snapshot.waitingList);
@@ -928,6 +997,19 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
      * Is the given task currently pending execution?
      */
     <span class="code-keyword">public</span> <span class="code-object">boolean</span> isPending(Task t) {
+<span class="code-keyword">if</span> (liveResultRequired() && lock.tryLock()) {
+    <span class="code-comment">// JENKINS-22708 and JENKINS-27871. Note we should get the lock always as it should be already held
</span>+    <span class="code-comment">// but the safer code construct is to get the lock anyway rather than assume it is held just by virtue
</span>+    <span class="code-comment">// of the current thread being the thread running maintain().
</span>+    <span class="code-keyword">try</span> {
+<span class="code-keyword">for</span> (BuildableItem i : pendings)
+    <span class="code-keyword">if</span> (i.task.equals(t))
+<span class="code-keyword">return</span> <span class="code-keyword">true</span>;
+<span class="code-keyword">return</span> <span class="code-keyword">false</span>;
+    } <span class="code-keyword">finally</span> {
+lock.unlock();
+    }
+}
 Snapshot snapshot = <span class="code-keyword">this</span>.snapshot;
 <span class="code-keyword">for</span> (BuildableItem i : snapshot.pendings)
     <span class="code-keyword">if</span> (i.task.equals(t))
@@ -1250,6 +1332,18 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
     }
 
     /**
+     * Some functionality that determines whether builds are blocked depending on what builds are currently running
+     * needs live information rather than information from the most recently updated snapshot. (See
+     * JENKINS-22708 and JENKINS-27871)
+     * @<span class="code-keyword">return</span> {@code <span class="code-keyword">true</span>} <span class="code-keyword">if</span> the current thread is running {@link #maintain()} and thus live results should be
+     * returned rather than results from {@link #snapshot}
+     * @since 1.FIXME
+     */
+    <span class="code-keyword">private</span> <span class="code-object">boolean</span> liveResultRequired() {
+<span class="code-keyword">return</span> <span class="code-object">Thread</span>.currentThread() == maintainThread;
+    }
+    
+    /**
      * Queue maintenance.
      *
      * <p>
@@ -1264,6 +1358,7 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
     <span class="code-keyword">public</span> void maintain() {
 lock.lock();
 <span class="code-keyword">try</span> { <span class="code-keyword">try</span> {
+    maintainThread = <span class="code-object">Thread</span>.currentThread();
 
     LOGGER.log(Level.FINE, <span class="code-quote">"Queue maintenance started {0}"</span>, <span class="code-keyword">this</span>);
 
@@ -1373,7 +1468,7 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>Queue <span class="code-keyword">extends</span> ResourceController <span class="code-keyword">implements</span> Saveable {
 <span class="code-keyword">else</span>
     LOGGER.log(Level.FINE, <span class="code-quote">"BuildableItem {0} with empty work units!?"</span>, p);
     }
-} <span class="code-keyword">finally</span> { updateSnapshot(); } } <span class="code-keyword">finally</span> {
+} <span class="code-keyword">finally</span> { maintainThread = <span class="code-keyword">null</span>; updateSnapshot(); } } <span class="code-keyword">finally</span> {
     lock.unlock();
 }
     }
</pre>
</div></div> 

<p>I am going to apply Occam's razor and apply a slightly tweaked version of the first fix to core. Jobs are not scheduled for execution at a particularly onerous rate such that the creation of a new snapshot should not cause undue GC pressure, plus the snapshot itself should be easy for GC to clean up and should not really ever end up being promoted out of the eden pool, so simplest fix works</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>Reassigned the issue to Stephen</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 core/src/main/java/hudson/model/Queue.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/5880ed830201f9349ae9def6653c19a186e1eb18" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/5880ed830201f9349ae9def6653c19a186e1eb18</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-27708&#93;</span><span class="error">&#91;FIXED JENKINS-27871&#93;</span>  Ensure that identification of blocked tasks is using the live state.</p>

<ul class="alternate" type="square">
	<li>The creation of a snapshot itself should be relatively cheap given the expected rate of<br/>
  job execution. You probably would need 100's of jobs starting execution every iteration<br/>
  of maintain() before this could even start to become an issue and likely the calculation<br/>
  of isBuildBlocked(p) will become a bottleneck before updateSnapshot() will. Additionally<br/>
  since the snapshot itself only ever has at most one reference originating outside of the stack<br/>
  it should remain in the eden space and thus be cheap to GC.</li>
</ul>


<ul class="alternate" type="square">
	<li><a href="https://issues.jenkins.io/browse/JENKINS-27708" title="Concurrent build limits not honored on Jenkins 1.607" class="issue-link" data-issue-key="JENKINS-27708"><del>JENKINS-27708</del></a> comments 225819 and 225906 provide more complex but logically equivalent fixes of<br/>
  this issue. I am favouring this approach as it is simpler and provides less scope for error as any<br/>
  new helper methods can just rely on the snapshot being up to date whereas with the other<br/>
  two candidates if a new helper method is introduced there is the potential to miss adding support<br/>
  for the live view. The comment 225819 has the risk of introducing extra lock contention while<br/>
  the comment 225906 version forces every access to the helper methods to pass a second memory<br/>
  barrier</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Stephen Connolly<br/>
Path:<br/>
 core/src/main/java/hudson/model/Queue.java<br/>
 test/src/test/java/hudson/model/QueueTest.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/152d00ad09931c10f02fab4ac8a42e574d622bd3" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/152d00ad09931c10f02fab4ac8a42e574d622bd3</a><br/>
Log:<br/>
  Merge pull request #1645 from stephenc/jenkins-27708</p>

<p><a href="https://issues.jenkins.io/browse/JENKINS-27708" title="Concurrent build limits not honored on Jenkins 1.607" class="issue-link" data-issue-key="JENKINS-27708"><del>JENKINS-27708</del></a>, <a href="https://issues.jenkins.io/browse/JENKINS-27871" title="Block on upstream projects does not work" class="issue-link" data-issue-key="JENKINS-27871"><del>JENKINS-27871</del></a>  Ensure that identification of blocked tasks is using the live state.</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/b688047877cf...152d00ad0993" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/b688047877cf...152d00ad0993</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dogfood">dogfood</a>:</i>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/blue.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/4087/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #4087</a></p>

<p>     Result = SUCCESS</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=alanbirtles">alanbirtles</a>:</i>
<p>Did this make it into 1.610?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p>Marking as fixed</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dvh_yxlon">dvh_yxlon</a>:</i>
<p>No, did not find a solution so far. I am going to create an issue here soon. Since we were still using an older version, I wanted to update first, check if the problem is still there and then come back to this issue.<br/>
Unfortunately the problem still exists, even after the update to version 1.613.</p>

<p>-> please notice <a href="https://issues.jenkins.io/browse/JENKINS-28376" title="jobs get stuck in queue from time to time" class="issue-link" data-issue-key="JENKINS-28376"><del>JENKINS-28376</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly">stephenconnolly</a>:</i>
<p>This is fundamentally an issue in t-c-b-p making incorrect assumptions</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=oleg_nenashev">oleg_nenashev</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=costescuandrei" class="user-hover" rel="costescuandrei">costescuandrei</a> , <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a>, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dvh_yxlon" class="user-hover" rel="dvh_yxlon">dvh_yxlon</a><br/>
Is it a subject for a new issue?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=costescuandrei">costescuandrei</a>:</i>
<p>Dirk already created <a href="https://issues.jenkins.io/browse/JENKINS-28376" title="jobs get stuck in queue from time to time" class="issue-link" data-issue-key="JENKINS-28376"><del>JENKINS-28376</del></a> for the new problem.<br/>
Don't know how related it is in code to the current one.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dogfood">dogfood</a>:</i>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/4292/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #4292</a><br/>
     <span class="error">&#91;FIXED JENKINS-27708&#93;</span><span class="error">&#91;FIXED JENKINS-27871&#93;</span>  Ensure that identification of blocked tasks is using the live state. (Revision 11b89d9c0b46f0afad0ca86a11264b503b799c2e)</p>

<p>     Result = UNSTABLE<br/>
ogondza : <a href="https://github.com/jenkinsci/jenkins/commit/11b89d9c0b46f0afad0ca86a11264b503b799c2e" class="external-link" target="_blank" rel="nofollow noopener">11b89d9c0b46f0afad0ca86a11264b503b799c2e</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/model/Queue.java</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-20211">JENKINS-20211</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-28084">JENKINS-28084</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-27650">JENKINS-27650</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-28887">JENKINS-28887</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-27871">JENKINS-27871</a>]</i>
